### PR TITLE
nativeaccess: try to load all located libsystemds

### LINKED ISSUE #107878: Elasticsearch fails to start with both 32 bit and 64 bit libsystemd installed
### Elasticsearch Version

8.15.0 snapshot

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

Ubuntu Linux

### Problem Description

Starting Elasticsearch immediately logs an error:
```
marshall@marshall-X570-AORUS-PRO-WIFI:~/dev/kibana$ yarn es snapshot -E xpack.security.authc.api_key.enabled=true --license trial
yarn run v1.22.19
$ node scripts/es snapshot -E xpack.security.authc.api_key.enabled=true --license trial
 info Installing from snapshot
 info version: 8.15.0
 info install path: /home/marshall/dev/kibana/.es/8.15.0
 info license: trial
 info Downloading snapshot manifest from https://storage.googleapis.com/kibana-ci-es-snapshots-daily/8.15.0/manifest-latest-verified.json
 info downloading artifact from https://storage.googleapis.com/kibana-ci-es-snapshots-daily/8.15.0/archives/20240422-131735_f9c414c5/elasticsearch-8.15.0-SNAPSHOT-linux-x86_64.tar.gz
 info downloading artifact checksum from https://storage.googleapis.com/kibana-ci-es-snapshots-daily/8.15.0/archives/20240422-131735_f9c414c5/elasticsearch-8.15.0-SNAPSHOT-linux-x86_64.tar.gz.sha512
 info checksum verified
 info extracting /home/marshall/dev/kibana/.es/cache/elasticsearch-8.15.0-SNAPSHOT-linux-x86_64.tar.gz
 info extracted to /home/marshall/dev/kibana/.es/8.15.0
 info created /home/marshall/dev/kibana/.es/8.15.0/ES_TMPDIR
 info setting secure setting bootstrap.password to changeme
 info Starting
   │ info bin/elasticsearch -E action.destructive_requires_name=true -E cluster.routing.allocation.disk.threshold_enabled=false -E ingest.geoip.downloader.enabled=false -E search.check_ccs_compatibility=true -E xpack.security.authc.api_key.enabled=true
   │ info ES_JAVA_OPTS: -Xms1536m -Xmx1536m
   │ERROR Apr 22, 2024 8:54:02 AM sun.util.locale.provider.LocaleProviderAdapter <clinit>
   │ERROR WARNING: COMPAT locale provider will be removed in a future release
   │ERROR OpenJDK 64-Bit Server VM warning: You have loaded library /usr/lib/i386-linux-gnu/libsystemd.so.0.26.0 which might have disabled stack guard. The VM will try to fix the stack guard now.
   │      It's highly recommended that you fix the library with 'execstack -c <libfile>', or link it with '-z noexecstack'.
   │ info [o.e.n.NativeAccess] [marshall-X570-AORUS-PRO-WIFI] Unable to load native provider. Native methods will be disabled. java.lang.UnsatisfiedLinkError: /usr/lib/i386-linux-gnu/libsystemd.so.0.26.0: /usr/lib/i386-linux-gnu/libsystemd.so.0.26.0: wrong ELF class: ELFCLASS32 (Possible cause: can't load IA 32 .so on a AMD 64 platform)
   │      	at java.base/jdk.internal.loader.NativeLibraries.load(Native Method)
   │      	at java.base/jdk.internal.loader.NativeLibraries$NativeLibraryImpl.open(NativeLibraries.java:331)
   │      	at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:197)
   │      	at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:139)
   │      	at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2418)
   │      	at java.base/java.lang.Runtime.load0(Runtime.java:852)
   │      	at java.base/java.lang.System.load(System.java:2025)
   │      	at org.elasticsearch.nativeaccess@8.15.0-SNAPSHOT/org.elasticsearch.nativeaccess.jdk.JdkSystemdLibrary.<clinit>(JdkSystemdLibrary.java:29)
   │      	at org.elasticsearch.nativeaccess@8.15.0-SNAPSHOT/org.elasticsearch.nativeaccess.lib.NativeLibraryProvider.getLibrary(NativeLibraryProvider.java:58)
   │      	at org.elasticsearch.nativeaccess@8.15.0-SNAPSHOT/org.elasticsearch.nativeaccess.LinuxNativeAccess.<init>(LinuxNativeAccess.java:20)
   │      	at org.elasticsearch.nativeaccess@8.15.0-SNAPSHOT/org.elasticsearch.nativeaccess.NativeAccessHolder.<clinit>(NativeAccessHolder.java:28)
   │      	at org.elasticsearch.nativeaccess@8.15.0-SNAPSHOT/org.elasticsearch.nativeaccess.NativeAccess.instance(NativeAccess.java:22)
   │      	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.initializeNatives(Elasticsearch.java:285)
   │      	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.initPhase2(Elasticsearch.java:169)
   │      	at org.elasticsearch.server@8.15.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:74)
   │ info [o.e.n.NativeAccess] [marshall-X570-AORUS-PRO-WIFI] Cannot check if running as root because native access is not available
```

This leads to a number of messages about `zstd is null` and shard failures, e.g.
```
│ info [o.e.c.r.a.AllocationService] [marshall-X570-AORUS-PRO-WIFI] failing shard [FailedShard[routingEntry=[.security-7][0], node[taa-iE1pTke-lCYe7gWuzA], [P], recovery_source[existing store recovery; bootstrap_history_uuid=false], s[INITIALIZING], a[id=aG-UKf6nRzW5RA4cjICx2w], unassigned_info[[reason=ALLOCATION_FAILED], at[2024-04-22T15:54:15.434Z], failed_attempts[1], delayed=false, last_node[taa-iE1pTke-lCYe7gWuzA], details[failed shard on node [taa-iE1pTke-lCYe7gWuzA]: shard failure, reason [refresh failed source[schedule]], failure java.lang.NullPointerException: Cannot invoke "org.elasticsearch.nativeaccess.Zstd.compressBound(int)" because "zstd" is null
```

And Elasticsearch crashes before finishing startup.

### Steps to Reproduce

Attempt to start Elasticsearch on a system with both 32 bit and 64 bit libsystemd installations.

```
marshall@marshall-X570-AORUS-PRO-WIFI:~$ sudo apt list | grep libsystemd

libsystemd-dev/eoan-security 242-7ubuntu3.11 amd64
libsystemd-dev/eoan-security 242-7ubuntu3.11 i386
libsystemd0/eoan-security,now 242-7ubuntu3.11 amd64 [installed,automatic]
libsystemd0/eoan-security,now 242-7ubuntu3.11 i386 [installed,automatic]
```

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by axw**
Linux systems with multiarch (e.g. i386 & x86_64) libraries may have libsystemd.0 in two subdirectories of an entry in java.library.path. For example, libsystemd.so.0 may be found in both /usr/lib/i386-linux-gnu and /usr/lib/x86_64-linux-gnu.

Instead of attempting to load any library found, attempt all and stop as soon as one is successfully loaded.

Fixes https://github.com/elastic/elasticsearch/issues/107878

### CODE REVIEW

File: libs/native/src/main21/java/org/elasticsearch/nativeaccess/jdk/JdkSystemdLibrary.java [hunk 1]
```diff
@@ -40,10 +44,23 @@ static String findLibSystemd() {
...
+                    .map(p -> p.toAbsolutePath().toString())
+                    .toList();
 
-                var foundpath = stream.filter(Files::isDirectory).map(p -> p.resolve(libsystemd)).filter(Files::exists).findAny();
-                if (foundpath.isPresent()) {
-                    return foundpath.get().toAbsolutePath().toString();
+                UnsatisfiedLinkError last = null;
+                for (String path : foundpaths) {
+                    try {
+                        System.load(path);
```
Comment (ldematte) [inline libs/native/src/main21/java/org/elasticsearch/nativeaccess/jdk/JdkSystemdLibrary.java:56]: A QQ to see if I got this correctly: in multiarch you may have multiple libraries in the library path, for different archs (32/64). Depending on which arch the JVM is running (64bit I assume, unless something really strange, but your solution should work either way), loading the lib for a different arch will fail with a UnsatisfiedLinkError.
So you load every library you find, the first one that succeeds will be the right arch. If we are not able to load any, throw as before. Correct?

Comment (axw) [reply libs/native/src/main21/java/org/elasticsearch/nativeaccess/jdk/JdkSystemdLibrary.java:56]: Correct :)

**Review by ldematte — APPROVED** [review]
Thanks! looks good to me, but I'd ask @rjernst for a second look.

**Review by rjernst — COMMENTED** [review]
The general pattern of trying multiple paths seems ok, but I would like to keep a separation of loading from identifying possible paths. Could we please keep the method findLibSystemd, but change it to return a List<String>? Then that can be looped over in the same static block that already tries to load.

**Comment by rjernst** [comment]
@axw One of test failures is due to tabs being present, can you please check indentation is with spaces?

**Comment by rjernst** [comment]
@elasticsearchmachine update branch

**Comment by axw** [comment]
:cry: so many different lint/check/spotless apply commands. Working on it, sorry.

**Comment by axw** [comment]
@rjernst all good now, sorry for that. PTAL.

**Review by rjernst — APPROVED** [review]
LGTM. Thanks @axw

**Pull request merged**